File: lab4/py/RouterPacket.py
#!/usr/bin/env python

from copy import deepcopy


class RouterPacket(object):
    def __init__(self, sourceID: int, destID: int, mincosts: list[int]):
        super(RouterPacket, self).__init__()
        # id of sending router sending this pkt
        self.sourceid: int = sourceID
        # id of router to which pkt being sent (must be an immediate neighbor)
        self.destid: int = destID
        # min cost to node 0 ... 3
        self.mincost: list[int] = deepcopy(mincosts)

    def clone(self):
        return RouterPacket(self.sourceid, self.destid, deepcopy(self.mincost))

File: lab4/cpp/src/RouterPacket.cpp
#include "RouterPacket.h"

#include <algorithm>
#include <vector>

using namespace std;

RouterPacket::RouterPacket(int sourceID, int destID, vector<int> mincosts)
    : sourceid{ sourceID }, destid{ destID }, mincost{ std::move(mincosts) } {}

/*
 * Clone packet information and allocate it on the heap to safetly pass it
 * from RouterNode::sendUpdate to RouterSimulator::evlist
 */
RouterPacket* RouterPacket::clone() const {
    return new RouterPacket{ sourceid, destid, vector<int>{ mincost } };
}
